### PR TITLE
ICU-23341 Support Gregorian era inheritance in Japanese calendar and …

### PR DESCRIPTION
**Pull request by yumaoka**
…DateFormat.

CLDR dropped Japanese calendar era data before Meiji. Also it introduced a new behavior that inheriting Gregorian calendar before Meiji (inheritEras: gregorian). This spec update requires following changes.

- DateFormatSmbols initialization code to load era data from calendar type specified by "inheritEras".
- Skip era index without definitions. (Japanese calendar inherit era 0 and 1 from Gregorian, but next era after 1 needs to be era 232 Meiji)

With these changes, this PR also removes `logKnownIssues` introduced by the previous CLDR data integration.

#### Checklist
- [x] Required: Issue filed: ICU-23341
- [x] Required: The PR title must be prefixed with a JIRA Issue number. Example: "ICU-NNNNN Fix xyz"
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. Example: "ICU-NNNNN Fix xyz"
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable
- [ ] Approver: Feel free to merge on my behalf

### CODE REVIEW

**Comment by markusicu** [comment]
Java: please run spotless

C++: lots of failures, mostly on Windows; maybe missing U_I18N_API somewhere?
```
erarulestest.obj : error LNK2019: unresolved external symbol "public: int __cdecl icu_79::EraRules::getNextEraCode(int)const " (?getNextEraCode@EraRules@icu_79@@QEBAHH@Z) referenced in function "private: void __cdecl EraRulesTest::testJapanese(void)" (?testJapanese@EraRulesTest@@AEAAXXZ) [D:\a\icu\icu\icu4c\source\test\intltest\intltest.vcxproj]
```

**Comment by yumaoka** [comment]
> Java: please run spotless

Reran spotless. I finished J version first and ran spotless. While working on C code, I want to make C and J code consistent as much as possible and modified J code.. and forgot to run spotless after that.

I think I need to install spotless plugin to IDE if available.

> C++: lots of failures, mostly on Windows; maybe missing U_I18N_API somewhere?

You're right. I missed U_I18N_API in a few places.

I updated the PR with the fixes. I also found integer 32 overflow warning in clang-ubsan build. This was caused by missing error status check before the reported location. I also fixed the issue. Anyway, I will check the updated CI build output.

**Review by richgillam — DISMISSED** [review]
I'm not sure I'm the best one to review this, but it all looks plausible, and it does seem to be passing all the tests now...

**Review by robertbastian — COMMENTED** [review]
tests lgtm

File: icu4j/main/core/src/main/java/com/ibm/icu/impl/EraRules.java [hunk 1]
```diff
@@ -206,7 +241,47 @@ public int getMaxEraCode() {
...
+     * @return Next era code. If not available, maximum era code is returned.
+     */
+    public int getNextEraCode(int eraCode) {
+        int nextEra = eraCode + 1;
+        if (nextEra < minEra) {
+            if (inheritEraRules != null) {
+                if (nextEra > inheritEraRules.getMaxEraCode()) {
+                    nextEra = minEra;
+                } else {
+                    nextEra = inheritEraRules.getNextEraCode(eraCode);
```
Comment (markusicu) [inline icu4j/main/core/src/main/java/com/ibm/icu/impl/EraRules.java:269]: It looks like we could get stuck here:

If nextEra == inheritEraRules.getMaxEraCode(), with nextEra = eraCode + 1, then we go back to the one-less eraCode and then to inheritEraRules.getNextEraCode(eraCode) which ... returns inheritEraRules.getMaxEraCode()?

Am I missing something?
Is this covered by unit tests?

File: icu4c/source/i18n/erarules.cpp [hunk 2]
```diff
@@ -103,22 +104,51 @@ static int32_t compareEncodedDateWithYMD(int encoded, int year, int month, int d
     }
 }
 
-EraRules::EraRules(LocalMemory<int32_t>& startDatesIn, int32_t startDatesLengthIn, int32_t minEraIn, int32_t numErasIn)
-    : startDatesLength(startDatesLengthIn), minEra(minEraIn), numEras(numErasIn) {
+EraRules::EraRules(const char* calTypeId, LocalMemory<int32_t>& startDatesIn, int32_t startDatesLengthIn,
+    int32_t minEraIn, int32_t numErasIn)
+    : startDatesLength(startDatesLengthIn), minEra(minEraIn), numEras(numErasIn), inheritEraRules(nullptr) {
+    uprv_strncpy(this->calTypeId, calTypeId, MAX_CAL_ID_LENGTH);
```
Comment (markusicu) [inline icu4c/source/i18n/erarules.cpp]: We should try to avoid strncpy().
Are these strings all either hardcoded, that is, known, or length-checked upstream?
If so, then we can assume that the ID fits (`uprv_strcpy()`); maybe U_ASSERT() that they do (strlen).
If not, then we should check that the ID fits; if yes: memcpy()+NUL; if no: buffer-overflow-error.

File: icu4c/source/i18n/erarules.cpp [hunk 3]
```diff
@@ -103,22 +104,51 @@ static int32_t compareEncodedDateWithYMD(int encoded, int year, int month, int d
     }
 }
 
-EraRules::EraRules(LocalMemory<int32_t>& startDatesIn, int32_t startDatesLengthIn, int32_t minEraIn, int32_t numErasIn)
-    : startDatesLength(startDatesLengthIn), minEra(minEraIn), numEras(numErasIn) {
+EraRules::EraRules(const char* calTypeId, LocalMemory<int32_t>& startDatesIn, int32_t startDatesLengthIn,
+    int32_t minEraIn, int32_t numErasIn)
+    : startDatesLength(startDatesLengthIn), minEra(minEraIn), numEras(numErasIn), inheritEraRules(nullptr) {
```
Comment (markusicu) [inline icu4c/source/i18n/erarules.cpp]:
```suggestion
EraRules::EraRules(const char* calTypeId, LocalMemory<int32_t>& startDatesIn, int32_t startDatesLengthIn,
        int32_t minEraIn, int32_t numErasIn)
        : startDatesLength(startDatesLengthIn), minEra(minEraIn), numEras(numErasIn), inheritEraRules(nullptr) {
```

File: icu4c/source/i18n/erarules.cpp [hunk 4]
```diff
@@ -103,22 +104,51 @@ static int32_t compareEncodedDateWithYMD(int encoded, int year, int month, int d
...
+    uprv_strncpy(this->calTypeId, calTypeId, MAX_CAL_ID_LENGTH);
+    this->calTypeId[MAX_CAL_ID_LENGTH] = 0;
     startDates = std::move(startDatesIn);
     initCurrentEra();
 }
 
 EraRules::~EraRules() {
+    if (inheritEraRules) {
```
Comment (markusicu) [inline icu4c/source/i18n/erarules.cpp]: optional: unnecessary check: delete nullptr is fine

File: icu4c/source/i18n/erarules.cpp [hunk 5]
```diff
@@ -257,19 +287,59 @@ EraRules* EraRules::createInstance(const char *calType, UBool includeTentativeEr
...
+        EraRules *tmpEraRules = createInstance(inheritCalType, includeTentativeEra, status);
+        if (tmpEraRules == nullptr) {
+            status = U_MEMORY_ALLOCATION_ERROR;
+            delete result;
+            return nullptr;
+        }
+        result->inheritEraRules = tmpEraRules;
+    }
+
     return result;
```
Comment (markusicu) [inline icu4c/source/i18n/erarules.cpp]: if U_FAILURE then createInstance() should delete result and return nullptr.

File: icu4c/source/i18n/erarules.cpp [hunk 6]
```diff
@@ -257,19 +287,59 @@ EraRules* EraRules::createInstance(const char *calType, UBool includeTentativeEr
     for (int32_t eraIdx = 0; eraIdx < eraStartDates.size(); eraIdx++) {
         startDates[eraIdx] = eraStartDates.elementAti(eraIdx);
     }
-    EraRules *result = new EraRules(startDates, eraStartDates.size(), minEra, numEras);
+    EraRules *result = new EraRules(calTypeId, startDates, eraStartDates.size(), minEra, numEras);
```
Comment (markusicu) [inline icu4c/source/i18n/erarules.cpp]: would be safer to use a LocalPointer, and use its constructor that takes an extra UErrorCode which then auto-sets the OOM

File: icu4c/source/i18n/dtfmtsym.cpp [hunk 7]
```diff
@@ -1957,48 +1982,77 @@ initField(UnicodeString **field, int32_t& length, CalendarDataSink &sink, CharSt
...
+
+    // Also load era names from resource corresponding to the inherited era rules if any.
+    const EraRules *inheritEraRules = eraRules->getInheritEraRules();
+    if (inheritEraRules != nullptr) {
+        loadEraNames(eras, length, eraNamesTable, calBaseBundle, eraWidth, inheritEraRules, status);
+    }
+}
+
+static void initEras(UnicodeString **field, int32_t& length, CalendarDataSink &sink, UResourceBundle *calBaseBundle,
+    const char *eraWidth, EraRules *eraRules, UErrorCode &status) {
```
Comment (markusicu) [inline icu4c/source/i18n/dtfmtsym.cpp]: Continuation line should be indented so that it doesn't look like the start of the function body.
```suggestion
        const char *eraWidth, EraRules *eraRules, UErrorCode &status) {
```

File: icu4c/source/i18n/dtfmtsym.cpp [hunk 8]
```diff
@@ -1957,48 +1982,77 @@ initField(UnicodeString **field, int32_t& length, CalendarDataSink &sink, CharSt
 }
 
 static void
-initEras(UnicodeString **field, int32_t& length, CalendarDataSink &sink, CharString &key, const UResourceBundle *ctebPtr, const char* eraWidth, int32_t maxEra, UErrorCode &status) {
-    if (U_SUCCESS(status)) {
-        length = 0;
-        UnicodeString keyUString(key.data(), -1, US_INV);
-        Hashtable *eraNamesTable = static_cast<Hashtable*>(sink.maps.get(keyUString));
+loadEraNames(UnicodeString *eras, int32_t length, Hashtable *eraNamesTable, UResourceBundle *calBaseBundle,
+    const char *eraWidth, const EraRules *eraRules, UErrorCode &status) {
```
Comment (markusicu) [inline icu4c/source/i18n/dtfmtsym.cpp]: Continuation line should be indented so that it doesn't look like the start of the function body.
```suggestion
        const char *eraWidth, const EraRules *eraRules, UErrorCode &status) {
```

**Review by markusicu — COMMENTED** [review]
- looks plausible
- small C++ issues
- I don't understand how the get-next-era code does not get into a loop

File: icu4j/main/core/src/main/java/com/ibm/icu/impl/EraRules.java (hunk 1 above)
Comment (yumaoka) [reply icu4j/main/core/src/main/java/com/ibm/icu/impl/EraRules.java:269]: The test case is below -

https://github.com/unicode-org/icu/pull/4019/changes#diff-dba4fd54aad245317e580f0d7d79902ea8b7993272e77d874edcc28245ee56b9R92

File: icu4c/source/i18n/erarules.cpp (hunk 2 above)
Comment (yumaoka) [reply icu4c/source/i18n/erarules.cpp]: >We should try to avoid strncpy().
OK, I'll make the change.

>Are these strings all either hardcoded, that is, known, or length-checked upstream?
If so, then we can assume that the ID fits (uprv_strcpy()); maybe U_ASSERT() that they do (strlen).
If not, then we should check that the ID fits; if yes: memcpy()+NUL; if no: buffer-overflow-error.

calTypeId is coming from CLDR and used as a key in calendar rule data. Because it's only coming from data managed by us, we should not really need error checking, although I'll put an assert.

File: icu4c/source/i18n/erarules.cpp (hunk 4 above)
Comment (yumaoka) [reply icu4c/source/i18n/erarules.cpp]: OK, I will updated.

File: icu4j/main/core/src/main/java/com/ibm/icu/impl/EraRules.java (hunk 1 above)
Comment (yumaoka) [reply icu4j/main/core/src/main/java/com/ibm/icu/impl/EraRules.java:269]: In case of Japanse calender

- EraRules has rule for Era 232 (Meiji - minimum era) to 236 (Reiwa - max era).
- EraRules has Gregorian EraRules as inheritEraRules
- Gregorian EraRules has two eras - 0 (BC - minmum era) and 1 (AD - max era).

getNextEraCode takes any arbitrary era code. The method add 1 to the input era code at the beginning.

- When eraCode is less than current EraRules minmum value (232 for Japanese calendar), checks with inheritEraRules (Gregorian).
  - If the value is greater than Gregorian's max era (1), then it returns current EraRule's minimum era. This is for jumping era Gregorian era 1 (AD) to Japanese era 232 (Meiji).
  - If the value is equal or less than Gregorian's max era (1), then it calls inheritEraRuls.getNextEra with the original input era code. Why does it do this? This design allows multiple EraRules inheritance (i.e. inheritEraRules may inherit another EraRules).

Comment (markusicu) [reply icu4j/main/core/src/main/java/com/ibm/icu/impl/EraRules.java:269]: Thanks... I think I got it now, and I think I got confused between the two layers.

File: icu4j/main/core/src/test/java/com/ibm/icu/dev/test/calendar/EraRulesTest.java [hunk 9]
```diff
@@ -78,5 +79,34 @@ public void testJapanese() {
...
+        // Note: Current CLDR data uses 1868-10-23 as the start date of Meiji.
+        // This is not really accurate, because it does not count Lunar-solar
+        // calendar system used before Meiji 6. This might be changed in future.
+        assertEquals("1868-10-23", JapaneseCalendar.MEIJI, rules.getEraCode(1868, 10, 23));
+        assertEquals("1868-10-22", GregorianCalendar.AD, rules.getEraCode(1868, 10, 22));
+        assertEquals("0001-01-01", GregorianCalendar.AD, rules.getEraCode(1, 1, 1));
+        assertEquals("0000-12-31", GregorianCalendar.BC, rules.getEraCode(0, 12, 31));
+        assertEquals("-1000-01-01", GregorianCalendar.BC, rules.getEraCode(-1000, 1, 1));
+
+        final int[] expectedEraCodes = {0, 1, 232, 233, 234, 235, 236};
```
Comment (markusicu) [inline icu4j/main/core/src/test/java/com/ibm/icu/dev/test/calendar/EraRulesTest.java:92]: How about also testing some weird inputs like -1, 2, 55, 231, 237, 238, 240, 555?

Comment (markusicu) [reply icu4j/main/core/src/test/java/com/ibm/icu/dev/test/calendar/EraRulesTest.java:92]: Actually, you already have a middle value and maxEra, but some more edge cases might be nice.

File: icu4c/source/i18n/erarules.cpp (hunk 2 above)
Comment (yumaoka) [reply icu4c/source/i18n/erarules.cpp]: Done

File: icu4c/source/i18n/erarules.cpp (hunk 3 above)
Comment (yumaoka) [reply icu4c/source/i18n/erarules.cpp]: Done

File: icu4c/source/i18n/erarules.cpp (hunk 4 above)
Comment (yumaoka) [reply icu4c/source/i18n/erarules.cpp]: Dropped if.

File: icu4c/source/i18n/erarules.cpp (hunk 5 above)
Comment (yumaoka) [reply icu4c/source/i18n/erarules.cpp]: Done along with LocalPointer change

File: icu4c/source/i18n/erarules.cpp (hunk 6 above)
Comment (yumaoka) [reply icu4c/source/i18n/erarules.cpp]: Changed to LocalPointer.

File: icu4c/source/i18n/dtfmtsym.cpp (hunk 7 above)
Comment (yumaoka) [reply icu4c/source/i18n/dtfmtsym.cpp]: Done

File: icu4c/source/i18n/dtfmtsym.cpp (hunk 8 above)
Comment (yumaoka) [reply icu4c/source/i18n/dtfmtsym.cpp]: Done

File: icu4j/main/core/src/test/java/com/ibm/icu/dev/test/calendar/EraRulesTest.java (hunk 9 above)
Comment (yumaoka) [reply icu4j/main/core/src/test/java/com/ibm/icu/dev/test/calendar/EraRulesTest.java:92]: Added a few more cases / minEra - 1 and maxEra + 1. These additional test cases should cover all conditions.

**Review by markusicu — APPROVED** [review]
thanks!

**Review by richgillam — APPROVED** [review]
Still LGTM.
